src/npcmove.cpp

```diff
@@ -795,77 +781,84 @@ npc_action npc::method_of_attack()
-                } else {
-                    return npc_aim;
-                }
-            }
-            if( !wont_hit_friend( tar ) ) {
-                if( dont_move ) {
-                    if( can_reload_current() ) {
-                        return npc_reload;
-                    } else {
-                        // Wait for clear shot
-                        return npc_pause;
-                    }
-                } else {
-                    return npc_avoid_friendly_fire;
-                }
-            } else if( !sees( *critter ) ) {
-                // Can't see target
-                return melee_action;
-            } else if( dist > confident && sees( tar ) ) {
-                // If out of confident range, aim or move closer to the target
-                if( dist > confident_gun_range( weapon, weapon.sight_dispersion( -1 ) ) ) {
-                    return melee_action;
-                } else {
-                    return npc_aim;
-                }
-            } else if( burst_size > 1 && dist <= confident / 3 &&
-                       weapon.ammo_remaining() >= burst_size &&
-                       (target_HP >= weapon.gun_damage() * 3 ||
-                        emergency( ai_cache.danger * 2 ) ) ) {
-                return npc_shoot_burst;
-            } else {
-                return npc_shoot;
+    // get any suitable modes excluding melee, any forbiden to NPCs and those without ammo
+    // if we require a silent weapon inappropriate modes are also removed
+    std::vector<std::pair<std::string,item::gun_mode>> modes;
+    if( rules.use_guns || !is_following() ) {
+        for( const auto &e : weapon.gun_all_modes() ) {
+            modes.push_back( e );
+        }
+
+        modes.erase( std::remove_if( modes.begin(), modes.end(),
+                     [&]( const std::pair<std::string, item::gun_mode>& e ) {
+
+            const auto &m = e.second;
+            return m.melee() || m.flags.count( "NPC_AVOID" ) ||
+                   !m->ammo_sufficient( m.qty ) || !can_use( *m.target ) ||
+                   ( rules.use_silent && is_following() && !m.target->is_silent() );
+
+        } ), modes.end() );
+    }
+
+    // modes outside confident range should always be the last option(s)
+    std::stable_sort( modes.begin(), modes.end(), [&]( const std::pair<std::string, item::gun_mode>& lhs,
+                                                       const std::pair<std::string, item::gun_mode>& rhs ) {
+        return ( confident_gun_range( lhs.second ) >= dist ) > ( confident_gun_range( rhs.second ) >= dist );
+    } );
+
+    // prefer modes that result in more total damage
+    std::stable_sort( modes.begin(), modes.end(), [&]( const std::pair<std::string, item::gun_mode>& lhs,
+                                                       const std::pair<std::string, item::gun_mode>& rhs ) {
+        return ( lhs.second->gun_damage() * lhs.second.qty ) > ( rhs.second->gun_damage() * rhs.second.qty );
+    } );
```

Comment:
You could try sorting once, with confident range being first priority and damage being second.

Though that won't be enough, since the real problem is NPCs wasting ammo on weak crap, not taking target difficulty nor hp into account, ignoring `emergency()`.